packages/ui-components/src/features/Form/FormInput.tsx

```diff
@@ -12,17 +12,35 @@ type HookFormInputWrapperProps = Record<string, unknown> & {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   Input: ComponentType<any>;
   required?: boolean;
+  nonWhitespace?: boolean;
+  validate?:
+    | ((value: string) => boolean | string)
+    | Record<string, (value: string) => boolean | string>;
 };
 export const FormInput = ({
   name,
   required,
+  nonWhitespace,
+  validate = {},
   options = {},
   Input,
   ...props
 }: HookFormInputWrapperProps) => {
   const { register, errors = {} } = useFormContext();
+
   const requiredConfig = required ? { required: 'Required' } : {};
-  const registerOptions = { ...options, ...requiredConfig };
+
+  const verifyIsNonwhitespace = (value: string) => !!value.trim() || 'Must not be empty';
```

Comment:
The separation between `reauired` and `nonWhitespace` is more about predictability. I (and Tom) also think that in *most* cases, it makes semantic sense to treat whitespace-only values as empty, but we thought it might be a bit bold to assume this is appropriate 100% of the time and just make it the default `FormInput` `required` behaviour.

The main (and so far only) edge case we can think of is that we probably don’t want to trim whitespace on password fields

Reply:
Hm, as is, the form shows `Required` when the field is empty and `Must not be empty` when it’s non-empty only whitespace. But perhaps that’s not a meaningful difference after all

Trimming by default except when `type="password"` sounds promising, though; I hadn’t thought of that. This might be the way to go

Reply:
So thinking about this, I actually think we don't want to do this much trimming, especially because this is inside the `ui-components` package and it gets used in multiple places. I suspect that it may end up causing more edge case bugs than it would solve.

So what I propose is instead, adding some logic when we check is the `required` error should show, to trim the whitespace then, but not actually trim it in the form input, just trim it for validation purposes.
E.g. instead of 
```required ? { required: 'Required' } : {};`

it might be something like:

```required ? { required: () => {
// check if the value with whitespace trimmed is not an empty string
// if it is, return 'Required'
// if not, return nothing (or whatever react-hook-form tells you to return when a field is valid)
} } : {};

Then if there is a one off case where we want to handle whitespace differently, we can add that into the `validate` property that one time, as opposed to adding it for everything

Reply:
I don’t have much control over when `options.required` does its thing (or when it does it), but I can make the `required` prop on `FormInput` a quasi-enum of type `boolean | 'non-whitespace'`. This gives us three levels of “required-ness”:

1. Not required
2. Required, but can be whitespace
3. Required, must not be whitespace

It should make `FormInput` nice to _use_, but its internal logic might be a bit unconventional. I’ll propose the change in this PR, and we can always revert if we see issues

Reply:
I expect it will be very rare cases when `options.required` is set, so you can ignore that case.

All this validation logic should be doing is:
- If a field is required, check if the trimmed value is of length > 0
- If it is, it is valid
- If it isn't, it is not valid and should show a message.
- 
We don't need anything more complicated than that I think, just because if we do anything more complex it is likely to cause some behaviour the user might not expect

Reply:
The trouble with that is validation would then always use the trimmed value, including on password fields. I’ve pushed a commit which gives the developer the choice between the three levels of required-ness, and it’s come out neater than I originally imagined

In terms of complexity, it’s equivalent to what you’ve suggested. Being just an extension, it also remains compatible with exiting/conventional usage of `<FormInput required />`